src/log_sifter.py

```diff
@@ -0,0 +1,161 @@
+                                        collect_statistics)
+        group_parser.parse_event_groups()
+        statistics = group_parser.statistics_summaries
+
+        # Construct the output
+        if self.flags.json_output:
+            self.output_to_json(group_parser, statistics)
+        else:
+            self.output_to_terminal(group_parser, statistics)
+
+    def output_to_terminal(self, group_parser, statistics):
+        """ Outputs a listing of each symptom burst and its details to
+            the terminal depending on the level of verbosity selected. """
+
+        burst_tags = list(group_parser.burst_dict.keys())
+        burst_dict = group_parser.burst_dict
+        burst_list = group_parser.bursts
+
+        symptom_list = group_parser.symptoms_found
+
+        if self.flags.verbosity == constants.VerbosityLevel.HIGH:
+            for symptom in symptom_list:
+                # come back to this after json rep is done.
+                pass
+        elif self.flags.verbosity == constants.VerbosityLevel.MEDIUM:
+            for burst in burst_list:
+                tag = burst.tag
+
+                print("\n{0} Burst".format(tag))
+                print("------------------------------------")
+                print(burst.burst_start_timestamp)
+                print("({0}) {1} symptoms detected.".format(burst.symptom_count, tag))
+                print("Burst duration: {0}".format(burst.duration))
+                print(burst.action_msg)
+                print("------------------------------------")
+
+        elif self.flags.verbosity == constants.VerbosityLevel.LOW:
+            for tag in burst_tags:
+                # Pull action message from first symptom in first burst
+                burst = burst_dict[tag][0]
+                print("\n{0}".format(tag))
+                print("------------------------------------")
+                print("{0} symptom detected.".format(tag))
+                print(burst.action_msg)
+                print("------------------------------------\n")
+            
+
+    def output_to_json(self, group_parser, statistics):
+        """ Unpacks the group parser into json output then writes the
+            results to gsyslyzeer_output.json. """
+
+        dict_output = self.convert_to_dict(group_parser)
+        with open("gsyslyzer_output.json", "w") as outfile:
+            json.dump(dict_output, outfile)
+
+    def convert_to_dict(self, object_to_translate):
+        """ Recursively unpacks objects stored within the gorup parser
+            to build JSON representations of the important pieces of
+            data that should be reported to the user from each nested
+            object. """
+
+        if isinstance(object_to_translate, list):
+            translated_list = []
+            for item in object_to_translate:
+                translated_list.append(self.convert_to_dict(item))
+
+            return translated_list
+        elif isinstance(object_to_translate, EventGroupParser):
```

Comment:
Using a big if-else tree based on types isn't really the proper way to write polymorphic code. If you need all these different objects to have a convert-to-dict operation then you should add such an operation to each one of these objects.

That's a bit tricky with list, because obviously you can't add a function to list. But at least reducing this code to a "check for a list, otherwise use the convert-to-dict function from the object" would be a major improvement.

Reply:
Ok, that makes more sense. I can break those out.

Reply:
I realized this effects some other objects not in this branch. For this change I'll need to spin off another PR where I actually add those convert self to dictionary functions, should that come before the resolution of this?